### PR TITLE
Zypper conveyorPacker

### LINKED ISSUE #1944: zypper bootstrap for singularity files has not been migrated from 2.6 to 3.0
Similar to #1942, the `zypper` bootstrap option has not been ported over yet.

### PR DESCRIPTION
**Pull request by ikaneshiro**
Conveyor Packer that allows for bootstrapping with zypper.

fixes #1944

### CODE REVIEW

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE file distributed with the sources of this project regarding your
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: .md

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+func (cp *ZypperConveyorPacker) Get(b *types.Bundle) (err error) {
+	cp.b = b
+
+	//check for zypper on system
+	zypperPath, err := exec.LookPath("zypper")
+	if err != nil {
+		return fmt.Errorf("zypper is not in PATH: %v", err)
+	}
+
+	// //check for rpm on system
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: double //

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+		osversion, ok = cp.b.Recipe.Header["osversion"]
+		if !ok {
+			return fmt.Errorf("Invalid zypper header, OSVersion referenced in mirror but no OSVersion specified")
+		}
+		mirrorurl = strings.Replace(mirrorurl, `%{OSVERSION}`, osversion, -1)
+	}
+
+	include, _ := cp.b.Recipe.Header["include"]
+
+	//check for include environment variable and add it to requires string
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: `//` + space

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+	include += ` ` + os.Getenv("INCLUDE")
+
+	//trim leading and trailing whitespace
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: `//` + space

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+	include = strings.TrimSpace(include)
+
+	//add aaa_base to start of include list by default
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: `//` + space

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+	include = `aaa_base ` + include
+
+	//Create the main portion of zypper config
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: `//` + space

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+	if err != nil {
+		return fmt.Errorf("While generating Zypper config: %v", err)
+	}
+
+	err = cp.copyPseudoDevices()
+	if err != nil {
+		return fmt.Errorf("While copying pseudo devices: %v", err)
+	}
+
+	//Add mirrorURL as repo
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: `//` + space

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+	cmd := exec.Command(zypperPath, `--root`, cp.b.Rootfs(), `ar`, mirrorurl, `repo-oss`)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err = cmd.Run(); err != nil {
+		return fmt.Errorf("While adding zypper mirror: %v", err)
+	}
+
+	//Refreshing gpg keys
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: `//` + space

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err = cmd.Run(); err != nil {
+		return fmt.Errorf("While refreshing gpg keys: %v", err)
+	}
+
+	args := []string{`--non-interactive`, `-c`, filepath.Join(cp.b.Rootfs(), zypperConf), `--root`, cp.b.Rootfs(), `--releasever=` + osversion, `-n`, `install`, `--auto-agree-with-licenses`}
+	args = append(args, strings.Fields(include)...)
+
+	//Zypper install command
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: `//` + space

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+	cmd = exec.Command(zypperPath, args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	sylog.Debugf("\n\tZypper Path: %s\n\tDetected Arch: %s\n\tOSVersion: %s\n\tMirrorURL: %s\n\tIncludes: %s\n", zypperPath, runtime.GOARCH, osversion, mirrorurl, include)
+
+	//run zypper
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: `//` + space

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+	} else if err != nil {
+		return fmt.Errorf("While bootstrapping from zypper: %v", err)
+	}
+
+	return nil
+}
+
+// Pack puts relevant objects in a Bundle!
+func (cp *ZypperConveyorPacker) Pack() (b *types.Bundle, err error) {
+
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: white line

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+	err = os.Chmod(cp.b.Rootfs()+"/.singularity.d/runscript", 0755)
+	if err != nil {
+		return
+	}
+
+	return nil
+}
+
+func (cp *ZypperConveyorPacker) genZypperConfig() (err error) {
+
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: white line

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+		return
+	}
+
+	//get mirrorURL, OSVerison, and Includes components to definition
+	mirrorurl, ok := cp.b.Recipe.Header["mirrorurl"]
+	if !ok {
+		return fmt.Errorf("Invalid zypper header, no MirrorURL specified")
+	}
+
+	//look for an OS version if the mirror specifies it
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: `//` + space

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+		return fmt.Errorf("zypper is not in PATH: %v", err)
+	}
+
+	// //check for rpm on system
+	err = rpmPathCheck()
+	if err != nil {
+		return
+	}
+
+	//get mirrorURL, OSVerison, and Includes components to definition
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: `//` + space

File: src/pkg/build/sources/conveyorPacker_zypper.go (hunk 2 above)
Comment (ArangoGutierrez) [reply src/pkg/build/sources/conveyorPacker_zypper.go]: and `//` + space

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]

File: src/pkg/build/sources/conveyorPacker_zypper_test.go [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+	"testing"
+
+	"github.com/sylabs/singularity/src/pkg/build/types"
+	"github.com/sylabs/singularity/src/pkg/test"
+)
+
+const zyppDef = "../../../../examples/opensuse/Singularity"
+
+func TestZypperConveyor(t *testing.T) {
+
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper_test.go]: white line

File: src/pkg/build/sources/conveyorPacker_zypper_test.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+	if testing.Short() {
+		t.SkipNow()
+	}
+
+	if _, err := exec.LookPath("zypper"); err != nil {
+		t.Skip("skipping test, zypper not found")
+	}
+
+	test.EnsurePrivilege(t)
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper_test.go]: `test.EnsurePrivilege(t)` should go on the first line of the func (niptick)

File: src/pkg/build/sources/conveyorPacker_zypper_test.go [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+	err = zc.Get(b)
+	//clean up tmpfs since assembler isnt called
+	defer os.RemoveAll(zc.b.Path)
+	if err != nil {
+		t.Fatalf("failed to Get from %s: %v\n", zyppDef, err)
+	}
+}
+
+func TestZypperPacker(t *testing.T) {
+
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper_test.go]: white line

File: src/pkg/build/sources/conveyorPacker_zypper_test.go [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+	if _, err := exec.LookPath("zypper"); err != nil {
+		t.Skip("skipping test, zypper not found")
+	}
+
+	test.EnsurePrivilege(t)
```
Comment (ArangoGutierrez) [inline src/pkg/build/sources/conveyorPacker_zypper_test.go]: same as above

**Review by ArangoGutierrez — DISMISSED** [review]

**Comment by ArangoGutierrez** [comment]
This PR LGTM, I don't have Open Suse to test it on dep

**Review by GodloveD — DISMISSED** [review]
I tried this on leap 42.3 and I got a panic.  Here's the stack trace.  Does this look like something I did or is there a bug?
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x9f6f99]

goroutine 1 [running]:
github.com/sylabs/singularity/src/pkg/build/sources.(*ZypperConveyorPacker).Get(0xc000482680, 0xc0000be640, 0x0, 0x0)
	/home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/sources/conveyorPacker_zypper.go:113 +0xbd9
github.com/sylabs/singularity/src/pkg/build.(*Build).Full(0xc0000be500, 0x1d, 0x7ffe614eee73)
	/home/vagrant/go/src/github.com/sylabs/singularity/src/pkg/build/build.go:156 +0xda
github.com/sylabs/singularity/src/cmd/singularity/cli.glob..func4(0x143c800, 0xc0004764c0, 0x2, 0x2)
	/home/vagrant/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli/build.go:106 +0x4d2
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).execute(0x143c800, 0xc0000300e0, 0x2, 0x2, 0x143c800, 0xc0000300e0)
	/home/vagrant/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:766 +0x2cc
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x143f9e0, 0xc0004c5f88, 0x4078c0, 0xc00003a0b8)
	/home/vagrant/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).Execute(0x143f9e0, 0x0, 0x0)
	/home/vagrant/go/src/github.com/sylabs/singularity/vendor/github.com/spf13/cobra/command.go:800 +0x2b
github.com/sylabs/singularity/src/cmd/singularity/cli.ExecuteSingularity()
	/home/vagrant/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli/singularity.go:100 +0x2d
main.main()
	/home/vagrant/go/src/github.com/sylabs/singularity/src/cmd/singularity/cli.go:16 +0x20
```

**Comment by ArangoGutierrez** [comment]
any update on this `WIP` ?

**Comment by ikaneshiro** [comment]
@ArangoGutierrez Not yet. Hoping to knock this out over the next day or two

**Comment by ikaneshiro** [comment]
The major issue encountered while developing and testing this feature was an exit 107 from Zypper due to the device-mapper package generating an error in its post script. This is fixed by bootstrapping with an openSUSE version of 42.3 or newer. I think this issue hasn't been encountered in past (to my knowledge) due to the 107 error code being a new addition to Zypper, meaning older versions of Zypper would not throw an error in this scenario. I suspect this is why bootstrapping with versions older than 42.3 appears to work on older openSUSE systems. As a result I have updated the openSUSE example definition to use 42.3 in order to avoid this error on systems using newer versions of Zypper

**Comment by ikaneshiro** [comment]
@ArangoGutierrez @jscook2345 @GodloveD @bauerm97 RFR

File: src/pkg/build/sources/conveyorPacker_zypper.go [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+	if rpmDBPath != `%{_var}/lib/rpm` {
+		return fmt.Errorf("RPM database is using a weird path: %s"+
+			"You are probably running this bootstrap on Debian or Ubuntu.\n"+
+			"There is a way to work around this problem:\n"+
+			"Create a file at path %s/.rpmmacros.\n"+
+			"Place the following lines into the '.rpmmacros' file:\n"+
+			"%s\n"+
+			"%s\n"+
+			"After creating the file, re-run the bootstrap.\n"+
+			"More info: https://github.com/singularityware/singularity/issues/241\n",
```
Comment (jscook2345) [inline src/pkg/build/sources/conveyorPacker_zypper.go]: sylabs/singularity

**Comment by jscook2345** [comment]
My tests pass on suse but it *requires* the special macros in /root/.rpmmacros. The verbose output is still pretty scary looking to. Things like 'root: user does not exist' etc are still present.

**Comment by ikaneshiro** [comment]
@jscook2345 sorry for the latency! 

After looking into the special macros issue I can't think of a better way to handle this ATM. Since there seems to be a lot of issues around the location of `_dbpath` being varied by package managers.

WRT the scary output. I don't know how much that can be fixed. It looks like using a newer version like 15.1  instread of 43.2 it doesn't have the `root: user does not exist` related warnings when I build on my VM. Maybe I should bump up the version in the example definition?

**Comment by jscook2345** [comment]
<pre>
FATAL:   While performing build: conveyor failed to get: RPM database is using a weird path: %{_usr}/lib/sysimage/rpmYou are probably running this bootstrap on Debian or Ubuntu.
</pre>

You're missing a newline here.

Can we also say 'non-standard path' location instead of 'weird path'. Also I think its OK to remove the comment about bootstrapping from other OSes.

**Comment by jscook2345** [comment]
<pre>
Warning: Checking for file conflicts requires not installed packages to be downloaded in advance in order to access their file lists. See option '--download-in-advance' in the zypper manual page for details.
</pre>

Perhaps we should enable / use this option?

**Comment by jscook2345** [comment]
I think after those last comments are fixed we can rebase and merge this assuming tests pass

**Comment by ikaneshiro** [comment]
Ill make the changes mentioned in both comments. I think having the check for file conflicts is a good idea

**Comment by ArangoGutierrez** [comment]
Any update here @ikaneshiro ?

**Comment by ikaneshiro** [comment]
@ArangoGutierrez So, I've been thinking about whether requiring a certain `.rpmmacros` configuration is necessary or not. I haven't been able explicitly find a case where it's needed, but I have a hunch that this will help prevent potential rpmdb incompatibilities between host and container package managers. This `.rpmmacros` requirement is also present in 2.x.

TLDR: I think this PR is ready to be merged once its passing CI

**Comment by ArangoGutierrez** [comment]
waaaahoooo
thanks for the update @ikaneshiro

File: internal/pkg/build/sources/conveyorPacker_zypper.go [hunk 20]
```diff
@@ -0,0 +1,244 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
```
Comment (ArangoGutierrez) [inline internal/pkg/build/sources/conveyorPacker_zypper.go]: 2018-2019

File: internal/pkg/build/sources/conveyorPacker_zypper_test.go [hunk 21]
```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
```
Comment (ArangoGutierrez) [inline internal/pkg/build/sources/conveyorPacker_zypper_test.go]: 2018-2019

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]

File: internal/pkg/build/sources/conveyorPacker_zypper.go (hunk 20 above)
Comment (ikaneshiro) [reply internal/pkg/build/sources/conveyorPacker_zypper.go]: good catch

**Comment by ArangoGutierrez** [comment]
PR work , I will leave @GodloveD and @jmstover how to better handle 
```
There is a way to work around this problem:
Create a file at path /root/.rpmmacros.
Place the following lines into the '.rpmmacros' file:
%_var /var
%_dbpath %{_var}/lib/rpm
After creating the file, re-run the bootstrap.
```
via user documentation maybe?

**Review by ArangoGutierrez — APPROVED** [review]
LGTM
